python/tests/test_metadata.py

```diff
@@ -400,3 +401,422 @@ def decode(self, encoded):
+        self.round_trip(schema, {"arr": [5, 6, 7], "int": 5})
+
+    def test_missing_and_extra_property(self):
+        schema = {
+            "codec": "struct",
+            "type": "object",
+            "properties": {
+                "int": {"type": "number", "binaryFormat": "i"},
+                "float": {"type": "number", "binaryFormat": "d"},
+            },
+        }
+        with self.assertRaisesRegex(
+            exceptions.MetadataValidationError, "'int' is a required property"
+        ):
+            self.round_trip(schema, {"float": 5.5})
+        with self.assertRaisesRegex(
+            exceptions.MetadataValidationError, "Additional properties are not allowed"
+        ):
+            self.round_trip(
+                schema, {"float": 5.5, "int": 9, "extra": "I really shouldn't be here"}
+            )
+
+    def test_bad_schema_union_type(self):
+        schema = {"codec": "struct", "type": ["object", "number"], "binaryFormat": "d"}
+        with self.assertRaisesRegex(
+            exceptions.MetadataSchemaValidationError, "is not one of"
+        ):
+            metadata.MetadataSchema(schema)
+        schema = {
+            "codec": "struct",
+            "type": "object",
+            "properties": {"test": {"type": ["number", "string"], "binaryFormat": "d"}},
+        }
+        with self.assertRaisesRegex(
+            exceptions.MetadataSchemaValidationError, "is not one of"
+        ):
+            metadata.MetadataSchema(schema)
+
+    def test_bad_schema_hetrogeneous_array(self):
+        schema = {
+            "codec": "struct",
+            "type": "array",
+            "items": [{"type": "number"}, {"type": "string"}],
+        }
+        with self.assertRaisesRegex(
+            exceptions.MetadataSchemaValidationError, "is not of type 'object'"
+        ):
+            metadata.MetadataSchema(schema)
+        schema = {
+            "codec": "struct",
+            "type": "object",
+            "properties": {
+                "hetro_array": {
+                    "type": "array",
+                    "items": [{"type": "string"}, {"type": "number"}],
+                }
+            },
+        }
+        with self.assertRaisesRegex(
+            exceptions.MetadataSchemaValidationError, "is not of type 'object'"
+        ):
+            metadata.MetadataSchema(schema)
+
+    def test_bad_binary_format(self):
+        schema = {"codec": "struct", "type": "number", "binaryFormat": "int"}
+        with self.assertRaisesRegex(
+            exceptions.MetadataSchemaValidationError, "does not match"
+        ):
+            metadata.MetadataSchema(schema)
+
+    def test_missing_binary_format(self):
+        schema = {
+            "codec": "struct",
+            "type": "number",
+        }
+        with self.assertRaisesRegex(
+            exceptions.MetadataSchemaValidationError,
+            "number type must have binaryFormat set",
+        ):
+            metadata.MetadataSchema(schema)
```

Comment:
This is excellent! Some additional tests would be good

- It would be nice to have examples in which we compare the encoded value to the known bytes representation. For example, a ``c`` encoding of 1, should have bytes value ``b[1]``. We could just compare directly to the expected output of struct.pack, but also good to have the raw bytes in there for a few examples as a sanity check.
- Likewise, we want to check the expected encoding for a few more complicated examples. For instance, we want to make sure that we're definitely not putting in extra padding bytes when we have, say a "b" and "Q" value packed next to each other.
- What happens when the user tries to specify endianness? Nice to see an explicit test for this.
- What happens when we try to decode garbage? Any potential trip hazards here?
- We want to capture the truncation of long strings behaviour and test for this. What happens when someone tries to encode too large an integer?

I think it'd be good to split things up into a few test classes here. What you have are mostly ``TestStructRoundTrip``, and we want to also have ``TestStructInterface`` and ``TestStructEncoding`` as well, I guess.

Reply:
Ah, also: strings. Are these strictly utf8? Can we put in unicode in there? How is the encoding decided? Is the length of the field the number of bytes, or the encoded width of the string? Need some tests around this.

Reply:
I've added quite a few tests for strings, alongside adding the `stringEncoding` and `nullTerminated` parameters.